app/models/registration.rb

```diff
@@ -4,13 +4,70 @@ class Registration < ApplicationRecord
   # Professors have extra privileges over assistants.
   enum role: [:student, :grader, :assistant, :professor]
 
+  validates_presence_of :course_id, :user_id
+  validates_inclusion_of :show_in_lists, in: [false, true]
+  
   # A registration is join model between a user and a course.
   belongs_to :user
   belongs_to :course
 
   has_one :term, through: :course
-  belongs_to :section, :primary_key => "crn"
+  has_many :registration_sections, dependent: :destroy
+  has_many :sections, through: :registration_sections
 
+  Section::types.each do |t, value|
+    # This reader is for convenience
+    define_method "#{t}_sections".to_sym do
+      self.sections.where(type: value)
+    end
+  end
+
+  attr_accessor :username
+  attr_accessor :orig_sections
```

Comment:
IIRC, has_many :through didn't work (it *is* declared, on line 16) to get this fully right.  It's why I have a save_sections method separate from a before_update or before_save callback; some attribute just wasn't getting set early enough or in a deterministic enough way, and so the validations kept failing.  Very frustration.  Much workaround.